react-redux/react-redux.d.ts

```diff
@@ -14,51 +14,72 @@ declare namespace ReactRedux {
   type Dispatch<S> = Redux.Dispatch<S>;
   type ActionCreator<A> = Redux.ActionCreator<A>;
 
-  interface ComponentDecorator<TOriginalProps, TOwnProps> {
-    (component: ComponentClass<TOriginalProps>|StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+
+  interface ComponentConstructor<P, S> extends __React.ComponentLifecycle<P, S> {
+  new(props?: P, context?: S) : void;
+  }
+
+  interface ComponentClassDecorator<TOriginalProps, TOwnProps> {
+  (component: ComponentConstructor<TOriginalProps, any> | ComponentClass<TOriginalProps> | StatelessComponent<TOriginalProps>): ComponentClass<TOwnProps>;
+  }
+
+  interface ComponentDecorator<TOriginalProps, TOwnProps, TState> {
```

Comment:
So I see that the magic here seems to be that there is a third type parameter. Does this imply that every usage of `connect` has to be parameterized to ensure that it typechecks correctly?